feat(gatsby-image): Art Direction SSR fix (compliments hydration fix PR)

polarathene opened the pull request:
## Description

Complimentary PR to [`gatsby-image` hydration fix](https://github.com/gatsbyjs/gatsby/pull/26097), this larger PR fixes the issue prior to React being available while loading HTML. 

If additional details about lines are needed, check individual commits for extra details or logical grouping.

## Notes

#### Potential changes up for discussion
As media queries cannot be inline styles, `<style>` elements are used within `<body>`(unless there is an API to add to `<head>`?), or rather each `gatsby-image` component instance. These are removed once React hydrates, so only valid in SSR output. Could change that if there's value in keeping around in the client?

The hash could also be used for cache keys instead of `img.src` if any value in that. 

#### Hash function implementation

The hash method DJB2a which is the same as what `styled-components` uses should be fine, there isn't any expectation for the results to be compatible with other DJB2a implementations, which might happen if the filepath strings contain chars that exceed a single byte(some languages, emoji, etc), that can be supported if desired but requires more code and adds perf overhead.

Further related details in my original PR comments:

- Originally considered/implemented base64 and FNV1a-32 before DJB2a for short CSS classes: https://github.com/gatsbyjs/gatsby/pull/24811#issuecomment-654543710
- DJB2a implementation notes, benchmark tests, etc: https://github.com/gatsbyjs/gatsby/pull/24811#issuecomment-654589043

#### `<noscript>` support not added

<details>
<summary>Old comments</summary>

This has not presently catered for `<noscript>`, which if desired, I'll revert to a previous `ResponsiveQueries` components in my original PR that used string concatenation with all CSS in a single `<style>` instead of one `<style>` per media condition + fallback/base. 

There is presently a PR actively discussing if `<noscript>` logic can be changed to drop the string variants supporting it without bumping the min React to 16.5+, or we can alternatively hold off until a v3 with breaking changes. 

Let me know if catering to `<noscript>` is presently worth the time(it's already partly out of sync IIRC).
</details>

**EDIT:** Shouldn't actually be a problem since this is only relevant for the SSR payload, and targets the root tag of the component.

#### Not exactly HTML spec compliant, but other popular packages use same approach

While browsers support this approach(and that libraries like Emotion apparently use this technique), `<style>` elements within body is technically against spec, and thus may be flagged if anyone has some CI / validation checks on HTML spec conformance. 

There's not much we can do about this without some SSR API improvements, or adopting some CSS-in-JS library.

Prior comment in original PR about `<style>` approach considerations: https://github.com/gatsbyjs/gatsby/pull/24811#issuecomment-643844535

## Related Issues

Related: https://github.com/gatsbyjs/gatsby/issues/24748#issuecomment-650181549

> While this kinda works, it results in a flash of the wrong image until the JS kicks in. If the `paddingBottom` was determined via media queries instead that would eliminate the issue.

Related: https://github.com/gatsbyjs/gatsby/pull/24811 (Original PR)

pvdz commented:
(Please don't remove the `status: triage needed` label. That just runs the risk of a PR being ignored because people thing "somebody else is handling it")

polarathene commented:
> That just runs the risk of a PR being ignored because people thing "somebody else is handling it")

Oh? I usually see someone come in and apply the tags like so, I assumed it was only for categorizing the PR/issue so took care of it in advance. 

As for handling it, you had requested that I reach out to you for the reviews, is there something else beyond that?

pvdz commented:
Those somebodies are usually Gatsby employees. For system parts that have no particular code owner, they take a somewhat light version of implied ownership (I mean, that's not set in stone but better than nothing). Just let us handle that label, please :)

I don't think there's anything to do for you in this PR right now. I have other prio's today so don't expect much from me until tomorrow.

polarathene commented:
Didn't realize hiding comments also hides/removes the reactions.. I had :+1: in understanding/agreement, but marked these comments as off-topic as they're unrelated to the actual review.

I'll be sure not to touch the labels for triage in future, thanks for clearing that up.

polarathene commented on packages/gatsby-image/src/index.js:
```diff
@@ -474,13 +529,16 @@ class Image extends React.Component {
-    if (fluid) {
-      const imageVariants = fluid
-      const image = getCurrentSrcData(fluid)
+    const image = getCurrentSrcData(imageVariants)
```
```suggestion
    // Initial client render state needs to match SSR until hydration finishes.
    // Once hydration completes, render again to update to the correct image.
    // `imageVariants` is always an Array type at this point due to `convertProps()`
    const image = !this.state.isHydrated
      ? imageVariants[0]
      : getCurrentSrcData(imageVariants)
```

polarathene commented on packages/gatsby-image/src/index.js:
```diff
@@ -443,6 +493,11 @@ class Image extends React.Component {
       draggable,
     } = convertProps(this.props)
 
+    const imageVariants = fluid || fixed
+    if (!imageVariants) {
+      return null
+    }
```
```suggestion
    const imageVariants = fluid || fixed
    if (!imageVariants) {
      // Abort early if missing image data (#25371)
      return null
    }
```

polarathene reviewed (commented):
Some of the changes from [this commit](https://github.com/gatsbyjs/gatsby/pull/26117/commits/d9b6c60e169b433b1700fa6fb250442906c5fb72), were added from this PR to the main [hydration fix PR](https://github.com/gatsbyjs/gatsby/pull/26097) ([relevant commit](https://github.com/gatsbyjs/gatsby/pull/26097/commits/a06a07b021683d7165f707f86ec9809bfea3576c)).

These changes would be relevant for rebasing, but have not been added due to the reliance on `isHydrated`) which this PR does not presently have any interaction with.

wardpeet commented:
Hey @polarathene, Thanks for doing all this work! It's working and the changes are dope but I don't think it's a good idea to accept these changes. Gatsby-image is already super complex and this makes it even more complex, we should let the user create these classes themselves. They can integrate it better in their CSS story (emotion, styled-components, css-modules, ...). Maybe we should just warn people about it in develop.

We never thought art-direction through and now we're here. Sorry for the time you've spent on it.

I don't want to close it just yet, so open for feedback.

polarathene commented:
Rebased onto master, extracted the feature out to a separate file as a HOC and removed any changes that could only be used within the `gatsby-image` component internally. That is explained with notes on using the HOC in [this commit message](https://github.com/gatsbyjs/gatsby/pull/26117/commits/8cefcbf51f62e54c81ac0c762ab5f0b94478a369).

You can reject this PR if you like. I updated the PR with an example HOC for the benefit of anyone that wants the feature but is unsure how to implement it. I have not taken time to test these changes. As it's not a class based component with constructor and has not been modified to use hooks, the renders will perform unnecessary overhead re-calculating values that have not changed.

Users on mobile or with fast connections may not experience the issue this PR solves. Might be helpful to link to this PR from the `gatsby-image` docs Art Direction section with a note about it being required for image dimensions to be correct with Art Direction prior to React being ready.

MaxwellKendall commented:
This needs to go into the gatsby-image documentation. Excellent work!

MaxwellKendall commented:
@polarathene I can open a PR for the dox to include this as a solution until the next delivery includes this out of the box.

polarathene commented:
@MaxwellKendall thanks that would be appreciated! :+1:

wardpeet commented:
Closing, this is fixed in our new [gatsby-plugin-image](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-image) component. Thank you for your contribution! <3
